File: model/src/main/java/io/sparkled/model/entity/SequenceChannel.java
package io.sparkled.model.entity;

import io.sparkled.model.animation.SequenceChannelEffects;

import javax.persistence.*;
import java.util.Objects;
import java.util.UUID;

@Entity
@Table(name = "sequence_channel")
public class SequenceChannel {

    @Id
    @Column(name = "uuid", nullable = false)
    private UUID uuid;

    @Basic
    @Column(name = "sequence_id", nullable = false)
    private int sequenceId;

    @Basic
    @Column(name = "stage_prop_uuid", nullable = false)
    private UUID stagePropUuid;

    @Basic
    @Column(name = "display_order", nullable = false)
    private Integer displayOrder;

    @Basic
    @Column(name = "channel_json", columnDefinition = "text")
    private String channelJson;

    public UUID getUuid() {
        return uuid;
    }

    public SequenceChannel setUuid(UUID uuid) {
        this.uuid = uuid;
        return this;
    }

    public int getSequenceId() {
        return sequenceId;
    }

    public SequenceChannel setSequenceId(int sequenceId) {
        this.sequenceId = sequenceId;
        return this;
    }

    public UUID getStagePropUuid() {
        return stagePropUuid;
    }

    public SequenceChannel setStagePropUuid(UUID stagePropUuid) {
        this.stagePropUuid = stagePropUuid;
        return this;
    }

    public Integer getDisplayOrder() {
        return displayOrder;
    }

    public SequenceChannel setDisplayOrder(Integer displayOrder) {
        this.displayOrder = displayOrder;
        return this;
    }

    /**
     * @return a {@link SequenceChannelEffects} instance in JSON form
     */
    public String getChannelJson() {
        return channelJson;
    }

    /**
     * @param channelJson a {@link SequenceChannelEffects} instance in JSON form
     */
    public SequenceChannel setChannelJson(String channelJson) {
        this.channelJson = channelJson;
        return this;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        SequenceChannel that = (SequenceChannel) o;
        return sequenceId == that.sequenceId &&
                Objects.equals(uuid, that.uuid) &&
                Objects.equals(stagePropUuid, that.stagePropUuid) &&
                Objects.equals(displayOrder, that.displayOrder) &&
                Objects.equals(channelJson, that.channelJson);
    }

    @Override
    public int hashCode() {
        return Objects.hash(uuid, sequenceId, stagePropUuid, displayOrder, channelJson);
    }
}

File: model/src/main/kotlin/io/sparkled/model/entity/SequenceChannel.kt
package io.sparkled.model.entity

import io.sparkled.model.animation.SequenceChannelEffects
import java.util.*
import javax.persistence.*

@Entity
@Table(name = "sequence_channel")
class SequenceChannel {

    @Id
    @Column(name = "uuid")
    private var uuid: UUID? = null

    @Basic
    @Column(name = "sequence_id")
    private var sequenceId: Int? = null

    @Basic
    @Column(name = "stage_prop_uuid")
    private var stagePropUuid: UUID? = null

    @Basic
    @Column(name = "name")
    private var name: String? = null

    @Basic
    @Column(name = "display_order")
    private var displayOrder: Int? = null

    @Basic
    @Column(name = "channel_json", columnDefinition = "text")
    private var channelJson: String? = null

    fun getUuid(): UUID? {
        return uuid
    }

    fun setUuid(uuid: UUID?): SequenceChannel {
        this.uuid = uuid
        return this
    }

    fun getSequenceId(): Int? {
        return sequenceId
    }

    fun setSequenceId(sequenceId: Int?): SequenceChannel {
        this.sequenceId = sequenceId
        return this
    }

    fun getStagePropUuid(): UUID? {
        return stagePropUuid
    }

    fun setStagePropUuid(stagePropUuid: UUID?): SequenceChannel {
        this.stagePropUuid = stagePropUuid
        return this
    }

    fun getName(): String? {
        return name
    }

    fun setName(name: String?): SequenceChannel {
        this.name = name
        return this
    }

    fun getDisplayOrder(): Int? {
        return displayOrder
    }

    fun setDisplayOrder(displayOrder: Int?): SequenceChannel {
        this.displayOrder = displayOrder
        return this
    }

    /**
     * @return a [SequenceChannelEffects] instance in JSON form
     */
    fun getChannelJson(): String? {
        return channelJson
    }

    /**
     * @param channelJson a [SequenceChannelEffects] instance in JSON form
     */
    fun setChannelJson(channelJson: String?): SequenceChannel {
        this.channelJson = channelJson
        return this
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as SequenceChannel

        if (uuid != other.uuid) return false
        if (sequenceId != other.sequenceId) return false
        if (stagePropUuid != other.stagePropUuid) return false
        if (name != other.name) return false
        if (displayOrder != other.displayOrder) return false
        if (channelJson != other.channelJson) return false

        return true
    }

    override fun hashCode(): Int {
        var result = uuid?.hashCode() ?: 0
        result = 31 * result + (sequenceId ?: 0)
        result = 31 * result + (stagePropUuid?.hashCode() ?: 0)
        result = 31 * result + (name?.hashCode() ?: 0)
        result = 31 * result + (displayOrder ?: 0)
        result = 31 * result + (channelJson?.hashCode() ?: 0)
        return result
    }

    override fun toString(): String {
        return "SequenceChannel(uuid=$uuid, sequenceId=$sequenceId, stagePropUuid=$stagePropUuid, name=$name, displayOrder=$displayOrder, channelJson=$channelJson)"
    }
}
